Добавляем анимацию появления шапки

Nintendodias opened the pull request:
---
:mortar_board: [Добавляем анимацию появления шапки](https://up.htmlacademy.ru/animation/1/user/1710269/tasks/3)

:boom: https://htmlacademy-animation.github.io/1710269-magic-vacation-1/3/

Nintendodias committed:
Добавляем анимацию появления шапки

dubaua commented on source/scss/animation/blocks.scss:
```diff
@@ -1,8 +1,10 @@
+.animation {
+  opacity: 0;
+  animation-fill-mode: both;
+  animation-timing-function: cubic-bezier(0,1.09,1,1);
+}
```
В задании указано использовать свойство `transition`, а не `animation`. `animation` лучше применять в случаях с зацикленной анимацией или когда анимация сложная и нам нужны `keyframes`.

Я рекомендую описать начальное состояние в селекторе `.page-header__logo` и подобных, а конечное состояние в каскаде `.load .page-header__logo`.

Nintendodias replied:
Здравствуйте!
Я написала специальный скрипт и миксин для работы с анимацией появления и выплывания. Зачем мне дублировать и повторять анимацию, переписывая на transition, если мне достаточно добавить класс и 2 data атрибута для этого? 
Предложенный вами способ только увеличивает и осложняет scss, что может сказаться на оптимизации работы сайта.

dubaua replied:
В задании мы хотим научиться пользоваться именно свойством `transition`, так как наша анимация проигрывается один раз и не требует сложных `keyframes`. 
Я посмотрел на работу миксина, но он неуместен в данном случае. В этом задании мы можем обойтись исключительно стилями и скриптом, который добавляет класс при загрузке страницы к `body`.
Ваш миксин подойдет для случаев, когда анимация конфигурируется на бекенде и приходит вместе с HTML в дата атрибутах.

Это хорошо, что вы понимаете как работают миксины. На небольших проектах миксины усложняют чтение и лучше написать пару одинаковых строк кода в нескольких местах, чем заводить миксин.

dubaua commented on source/scss/animation/blocks.scss:
```diff
@@ -1,8 +1,10 @@
-    opacity: 0;
-    animation-fill-mode: both;
-
     &._type {
       &_fadeIn {
         animation-name: fadeIn;
```
Настоятельно не рекомендую использовать такие конструкции для склеивания модификаторов. Это усложняет поиск стилей по проекту. Например, мы увидели в разметке селектор `_type_slideInLeft` и попытаемся найти его в стилях, то не найдем, потому что он склеивается из двух кусочков. Другой разработчик может не догадаться об этом, так как такой подход выходит за рамки методологии БЭМ.

Также, в данном случае, не рекомендую использовать миксин. В данном случае можно продублировать повторяющиеся правила стилей для логотипа, меню и футера. Это займет чуть больше строк кода, но будет проще для чтения.

dubaua commented on source/scss/animation/menu.scss:
```diff
@@ -1,29 +1,31 @@
-      &.active {
+.load {
+  header {
+    .js-menu-link:after {
```
Селекторы с префиксом `js-` используются для выборки элементов в скриптах и, по-хорошему, не должны использоваться для стилизации. Если нужно стилизовать ссылку, лучше завести для этого свой класс.

dubaua commented on source/scss/animation/page_main.scss:
```diff
@@ -0,0 +1,26 @@
+}
+.load {
+  .screen__footer {
+    transform: translateY(0%)
```
Здесь задание выполнено верно, молодец!

dubaua commented on source/scss/animation/page_main.scss:
```diff
@@ -0,0 +1,26 @@
+.screen__footer {
+    transform: translateY(100%);
+    transition: 1s;
```
Лучше не отдавать браузеру самому заполнять неуказанные свойства для сокращенной записи `transition`. Советую писать полностью необходимые свойства в следующем порядке:
1. transition-property;
2. transition-duration;
3. transition-timing-function;
4. transition-delay,
либо явно указывать каждое правило отдельно.

dubaua commented on source/scss/animation/menu.scss:
```diff
@@ -1,29 +1,31 @@
+    .js-menu-link:after {
+      opacity: 0;
+    }
+    &.active .js-menu-link:after, .js-menu-link.active:after  {
```
Если правила создаются для нескольких селекторов, то лучше каждый из них писать с новой строки. Иначе, при беглом просмотре может показаться, что это один большой селектор.

dubaua commented on source/scss/animation/page_main.scss:
```diff
@@ -0,0 +1,26 @@
+    @include animation;
+  }
+  .social-block__toggler::before {
+    animation: start cubic-bezier(0,1.09,1,1) .5s 1s both;
```
В этом случае тоже можно обойтись свойством `transition` для `transform: scale` и применить кривую Безье с узлами большими единицы для того, чтобы фоновый кружок увеличивался в середине анимации больше, чем его конечный размер. Попробуй `cubic-bezier(0, 0, 0.5, 1.5)`

Nintendodias replied:
Не очень понимаю, как с помощью transition вернуть состояние от:
1 => 1.1 (можно сделать только с transition)
1.1 => 1 (это уже комбинированная анимация, потому я и подключила animation)

dubaua replied:
Смотри пример: https://codepen.io/dubaua/pen/KKgEJpj (нужно навести мышь на серый круг)
Анимация с помощью `transition` описывается начальным и конечным состоянием. Обрати внимание на кривую Безье на 20 строке: второй узел `0.5, 2`. Он позволяет изменить `scale` больше, чем на 1. 
Если посмотреть эту кривую в отладчике браузера, то получится 
![image](https://user-images.githubusercontent.com/1058952/105180988-0d60f700-5b4d-11eb-8b53-6b13db3ea8b0.png)
Таким образом, можно выходить за рамки начального и конечного состояния `transition`

dubaua reviewed (changes requested):
Нужна доработка в анимации логотипа, меню и круга соцсетей, а с футером все верно, используй его как пример.

Nintendodias committed:
fix

1. fix _type_fadeIn...
2. fix selectors starting js-...
3. fix all parameters for transition
4. fix selectors by common

dubaua reviewed (changes requested):
Сейчас анимация подчеркивания пункта меню "главная", кружка соцсетей и текста "ОТГАДАЙ ПЕРВЫМ МЕСТО..." запускается значительно позже остальных. Давай попробуем управлять анимацией с помощью одного подписчика `load` на `body` и отдельными правилами с `transition` для каждого элемента главной страницы? Последовательностью появления можно управлять с помощью `transition-delay`.

Если что-то непонятно, можем запланировать видеосозвон, я помогу разобраться.
